apps/web/src/components/costs/energy-projection-calculation.tsx

```diff
@@ -0,0 +1,201 @@
+}
+
+export function getPredictedCostForMonth(energyData: EnergyData[], userData: UserData[]): number {
+    if (userData.length === 0 || energyData.length === 0) {
+        return 0;
+    }
+
+    const today: Date = new Date();
+    const firstDayOfMonth: Date = new Date(today.getFullYear(), today.getMonth(), 1);
+    const lastDayOfMonth: Date = new Date(today.getFullYear(), today.getMonth() + 1, 0);
+    const daysPassed: number = Math.floor((today.getTime() - firstDayOfMonth.getTime()) / (24 * 60 * 60 * 1000)) + 1;
+    const daysInMonth: number = lastDayOfMonth.getDate();
+    const user = userData[1];
+
+    const totalConsumptionCurrentMonth = getCalculatedTotalConsumptionCurrentMonth(energyData);
+
+    const monthlyUsage: number = (totalConsumptionCurrentMonth / daysPassed) * daysInMonth;
+
+    return Number.parseFloat((monthlyUsage * user.workingPrice + user.basePrice).toFixed(2));
+}
+
+export function getPredictedCostForWeek(energyData: EnergyData[], userData: UserData[]): number {
+    if (userData.length === 0 || energyData.length === 0) {
+        return 0;
+    }
+
+    const today: Date = new Date();
+    const daysPassed: number = today.getDay() + 1;
+    const user = userData[1];
+
+    const totalConsumptionCurrentWeek = getCalculatedTotalConsumptionCurrentWeek(energyData);
+    const weeklyUsage: number = (totalConsumptionCurrentWeek / daysPassed) * 7;
+
+    const weekBasePrice = user.basePrice / 4;
+    return Number.parseFloat((weeklyUsage * user.workingPrice + weekBasePrice).toFixed(2));
+}
+
+export function getPredictedCostForDay(energyData: EnergyData[], userData: UserData[]): number {
+    if (userData.length === 0 || energyData.length === 0) {
+        return 0;
+    }
+
+    const user = userData[1];
+    const totalConsumptionCurrentDay = getCalculatedTotalConsumptionCurrentDay(energyData);
+    const currentTime = new Date();
+    const hoursPassed = currentTime.getHours() + currentTime.getMinutes() / 60;
+    const projectedDailyConsumption = (totalConsumptionCurrentDay / hoursPassed) * 24;
+    const daysInMonth = new Date(currentTime.getFullYear(), currentTime.getMonth() + 1, 0).getDate();
+    const dailyBasePrice = user.basePrice / daysInMonth;
+    const predictedCost = Number.parseFloat(
+        (projectedDailyConsumption * user.workingPrice + dailyBasePrice).toFixed(2),
+    );
+
+    return predictedCost;
+}
+
+// Vergleich zu anderem Tag/Woche/Monat
+
+function calculateComparison(current: number, previous: number) {
+    const absoluteDifference = current - previous;
+    const relativeDifference = previous !== 0 ? (absoluteDifference / previous) * 100 : 0;
+    return {
+        absoluteDifference: Number.parseFloat(absoluteDifference.toFixed(2)),
+        relativeDifference: Number.parseFloat(relativeDifference.toFixed(2)),
+    };
+}
+
+function getCalculatedTotalConsumptionPreviousDay(data: SensorDataSelectType[]): number {
+    const previousDate = new Date();
```

Comment:
Das ist zwar ein extremer Randfall, aber in den Methoden, die diese "privaten" Methoden aufrufen, hast du jeweils auch eine Variable mit dem aktuellen Tag. Wenn alles schlecht läuft, kann es passieren, dass diese Methode schon am nächsten Tag aufgerufen. Das gestern in dieser Methode wäre dann sogesehen das heute von der aufrufenden Methode :smile:

Reply:
Ich check gar nichts😂